exist-core/src/main/java/org/exist/xquery/XQueryContext.java

```diff
@@ -3379,27 +3239,23 @@ public void registerBinaryValueInstance(final BinaryValue binaryValue) {
     public static class BinaryValueCleanupTask implements CleanupTask {
         @Override
         public void cleanup(final XQueryContext context, final Predicate<Object> predicate) {
-            if (context.binaryValueInstances != null) {
-                List<BinaryValue> removable = null;
-                for (final BinaryValue bv : context.binaryValueInstances) {
-                    try {
-                        if (predicate.test(bv)) {
-                            bv.close();
-                            if (removable == null) {
-                                removable = new ArrayList<>();
-                            }
-                            removable.add(bv);
-                        }
-                    } catch (final IOException e) {
-                        LOG.error("Unable to close binary value: {}", e.getMessage(), e);
+            if (context.binaryValueInstances == null) {
+                return;
+            }
+            final List<BinaryValue> removable = new ArrayList<>();
```

Comment:
Another issue with a change to eager evaluation from lazy evaluation. If there are no binary value instances which is likely true for many many XQuery's then this function will use more memory and be less performant.